drake/core/Function.h

```diff
@@ -0,0 +1,148 @@
+
+#include <string>
+#include <vector>
+#include <stdexcept>
+#include <iostream>
+#include <Eigen/Dense>
+#include "Gradient.h"
+
+namespace Drake {
+
+  /** InputOutputRelation
+   * \brief Tags which can be used to inform algorithms about underlying structure in a function
+   * e.g., linear, affine, polynomial, analytic, differentiable, continuous, measurable, and -- lastly -- arbitrary
+   *
+   * note: i considered using inheritance to capture the relationship, but passing around types at runtime
+   * was more of a pain than simply capturing the inheritance with the helper functions below.
+   */
+  struct InputOutputRelation {
+    enum Form {
+      ARBITRARY,
+      DIFFERENTIABLE,
+      POLYNOMIAL,
+      AFFINE,
+      LINEAR,
+      CONSTANT,
+      ZERO
+      // todo: add more (rational, ...)
+    };
+    Form form;
+    // todo: add sparsity info
+
+    InputOutputRelation(Form f) : form(f) { };
+
+    static bool isa(const Form &f, const Form &base) {
```

Comment:
isA? Sticking to the code formatting rules is more important than using the same name as Matlab in my opinion.